deployment-server/src/Deploy/Worker.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings          #-}
+module Deploy.Worker where
+
+import           Control.Concurrent.Chan     (Chan, readChan)
+import           Control.Monad               (forever)
+import           Control.Monad.IO.Class      (liftIO)
+import           Control.Monad.Reader        (runReaderT)
+import           Control.Newtype.Generics    (Newtype, unpack)
+import qualified Data.ByteString.Lazy.Char8  as BS
+import           Data.Char                   (isSpace)
+import           Data.Text                   (Text)
+import qualified Data.Text                   as Text
+import           GHC.Generics                (Generic)
+import           GitHub.Data.Webhooks.Events (PullRequestEvent)
+import           Options.Generic             (ParseField, ParseFields, ParseRecord, getRecord)
+import           System.Exit                 (ExitCode (ExitFailure, ExitSuccess))
+import           System.IO.Temp              (withSystemTempDirectory)
+import           System.Process.Typed        (readProcess, runProcess, setEnv, setWorkingDir, shell)
+import           Web.Slack                   (SlackConfig, chatPostMessage)
+import           Web.Slack.Chat              (mkPostMsgReq)
+
+newtype SlackChannel = SlackChannel Text
+    deriving stock (Generic)
+    deriving anyclass (Newtype, ParseFields, ParseRecord, ParseField)
+    deriving newtype (Show)
+
+instance Read SlackChannel where
+    readsPrec _ s = [(SlackChannel . Text.pack $ s, "")]
+
+newtype Deployment = Deployment Text
+    deriving stock (Generic)
+    deriving anyclass (Newtype, ParseFields, ParseRecord, ParseField)
+    deriving newtype (Show)
+
+instance Read Deployment where
+    readsPrec _ s = [(Deployment . Text.pack $ s, "")]
+
+runWorker :: Chan PullRequestEvent -> FilePath -> FilePath -> [String] -> SlackConfig -> SlackChannel -> Deployment -> IO ()
+runWorker chan configDir stateFile extraIncludes slackConfig slackChannel deployment = forever $ do
+    event <- readChan chan
+    deploy event configDir stateFile extraIncludes slackConfig slackChannel deployment
+
+deploy :: PullRequestEvent -> FilePath -> FilePath -> [String] -> SlackConfig -> SlackChannel -> Deployment -> IO ()
+deploy event configDir stateFile extraIncludes slackConfig slackChannel deployment = withSystemTempDirectory "deployment" $ \tempDir -> liftIO $ do
+    let plutusDir = tempDir <> "/plutus"
+        nixopsDir = plutusDir <> "/deployment/nixops"
+    putStrLn "Deploy origin/master"
+    runIn tempDir "git clone https://github.com/input-output-hk/plutus.git"
+    runIn plutusDir "git checkout origin/master"
+    (_, stdout', _) <- readIn plutusDir "git rev-parse HEAD"
+    let gitHead = rstrip . BS.unpack $ stdout'
+        extraIncludesString = unwords . fmap (\s -> "-I" <> s) $ extraIncludes
+        args = extraIncludesString <+> "-s" <+> stateFile <+> "-d" <+> (Text.unpack . unpack) deployment
+    runIn nixopsDir $ "cp" <+> configDir <> "/*.json ."
```

Comment:
What happens if I set `configDir` to `"/etc/passwd ./the/real/config/dir/"`?

I think I'd be able to copy arbitrary files, as root, to a target directory.

Reply:
only json files

Reply:
Why only json files? I mean, I think I can maliciously use this code to create the command `cp /etc/passwd ./real/configdir/*.json .`.

Reply:
it's `cp` the other way round, the worst you could do is `cp /etc/passwd/*.json .`

Reply:
I think you're assuming that `configDir` doesn't have spaces in it.

Reply:
ah, sneaky. you would only be able to copy these files to the temporary directory that is created during execution. Ofc not good but you would have to also find a way to execute arbitrary code to do something with these files and if you can do that you don't need to copy these files because you can do anything you want anyway!

Reply:
Not true! If the deployment takes 5 minutes, that's plenty of time to run `find /tmp/ -type f -name passwd`, grab it, and take it to another box for processing.

(I admit we're under a larger game-over scenario here, but it seems a good time to point out some sneakier attack vectors that I know about.)

Reply:
Perhaps there's a way I can make sure all these files are only readable by `root`.

Reply:
If I `cp -p` that should be enough I think since relevant permissions will remain.

Reply:
ok, made `cp` the other way round so that `configDir` must be a directory otherwise `cp` will fail